### PR TITLE
[user-authn] multitenancy RBAC

### PR DESCRIPTION
**Pull request by yalosev**
Signed-off-by: Yuriy Losev <yuriy.losev@flant.com>

## Description
Use embedded RBAC instead of webhook auth. Pay attention: If you use limitNamespaces with regexp, you can't create namespaces in a Chart with other resources in this namespace because of kubernetes nature. You have to create namespace before, it will get RBAC rules and then you will be able to create other resources

## Why do we need it, and what problem does it solve?
Webhook couldn't handle security well. Because if webhook could be accessed - it passes everything

## Changelog entries

```changes
section: user-authz
type: feature
summary: Rework multitenancy ClusterAuthorizationRules, they work with namespaces RBAC now
impact: API-server will be restarted. If you use limitNamespaces with regexp, you can't create namespaces in a Chart with other resources in this namespace because of kubernetes nature. You have to create namespace before, it will get RBAC rules and then you will be able to create other resources
impact_level: high
```

### CODE REVIEW

**Comment by yalosev** [comment]
Please, check the solution but it has some unresolvable issues. I guess we should discuss them

File: ee/modules/140-user-authz/hooks/internal/cluster_auth_rule.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+*/
+
+package internal
+
+import (
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ClusterAuthorizationRule struct {
```
Comment (apolovov) [inline ee/modules/140-user-authz/hooks/internal/cluster_auth_rule.go:13]: Why don't you put this specs to `/modules/140-user-authz/hooks/internal`?

File: ee/modules/140-user-authz/hooks/multi_handle_rules.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2021 Flant JSC
```
Comment (apolovov) [inline ee/modules/140-user-authz/hooks/multi_handle_rules.go:2]:
```suggestion
Copyright 2022 Flant JSC
```

File: tools/build_includes/modules-EE.yaml [hunk 3]
```diff
@@ -9,8 +9,22 @@
   to: /deckhouse/modules/040-node-manager/cloud-providers/vsphere
 - add: /ee/modules/110-istio
   to: /deckhouse/modules/110-istio
-- add: /ee/modules/140-user-authz/templates/webhook
-  to: /deckhouse/modules/140-user-authz/templates/webhook
+- add: /ee/modules/140-user-authz/templates/multi-cluster-role-bindings.yaml
```
[the remark this answers is not in the document]
Comment (apolovov) [reply tools/build_includes/modules-EE.yaml:12]: It is because .build.yaml

File: ee/modules/140-user-authz/hooks/multi_handle_rules_test.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+			f.BindingContexts.Set(f.KubeStateSet(defaultNS + rule1))
+			f.RunHook()
+		})
+
+		It("`global.clusterIsBootstrapped` must not exist", func() {
+			Expect(f).To(ExecuteSuccessfully())
+			Expect(f.ValuesGet("global.clusterIsBootstrapped").Exists()).To(BeFalse())
+			fmt.Println(f.ValuesGet("userAuthz.internal.multitenancyCRDs"))
+		})
+
```
Comment (apolovov) [inline ee/modules/140-user-authz/hooks/multi_handle_rules_test.go:37]: it is strange that it is enough only one test case.

File: ee/modules/140-user-authz/hooks/multi_handle_rules_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	})
+})
+
+const (
```
Comment (apolovov) [inline ee/modules/140-user-authz/hooks/multi_handle_rules_test.go:41]: Why is declaration below code?

File: ee/modules/140-user-authz/templates/multitenancy.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{{- if .Values.userAuthz.enableMultiTenancy }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: d8-multitenancy
```
Comment (apolovov) [inline ee/modules/140-user-authz/templates/multitenancy.yaml:5]:
```suggestion
  name: multitenancy-is-enabled
```
It is our namespace, so we don't need d8 prefix.

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/140-user-authz/templates/multi-cluster-role-bindings.yaml [hunk 7]
```diff
@@ -0,0 +1,64 @@
+{{- range $crd := .Values.userAuthz.internal.multitenancyCRDs }}
+  {{- if $crd.spec.limitNamespaces }}
```
[the remark this answers is not in the document]
Comment (yalosev) [reply ee/modules/140-user-authz/templates/multi-cluster-role-bindings.yaml]: nope. Removed

File: ee/modules/140-user-authz/hooks/multi_handle_rules.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+		multitenancyRules = append(multitenancyRules, oRule)
+	}
+
+	input.Values.Set("userAuthz.internal.multitenancyCRDs", multitenancyRules)
+
+	return nil
+}
+
+type openAPIRule struct {
```
[the remark this answers is not in the document]
Comment (yalosev) [reply ee/modules/140-user-authz/hooks/multi_handle_rules.go]: renamed and moved to internal package

File: tools/build_includes/modules-EE.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (yalosev) [reply tools/build_includes/modules-EE.yaml]: moved to dir

File: ee/modules/140-user-authz/hooks/internal/cluster_auth_rule.go (hunk 1 above)
Comment (yalosev) [reply ee/modules/140-user-authz/hooks/internal/cluster_auth_rule.go]: removed

File: ee/modules/140-user-authz/hooks/multi_handle_rules.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
+/*
```
[the remark this answers is not in the document]
Comment (yalosev) [reply ee/modules/140-user-authz/hooks/multi_handle_rules.go]: moved to dir, prefix removed.

**Review by apolovov — APPROVED** [review]

**Review by apolovov — CHANGES_REQUESTED** [review]
Get rid of control-plane-configurator.

**Review by apolovov — APPROVED** [review]

File: ee/modules/140-user-authz/templates/multitenancy/role-bindings.yaml [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  {{- if hasKey $crd.spec "accessLevel" }}
+    {{- if not (list "User" "PrivilegedUser" "Editor" "Admin" "ClusterEditor" "ClusterAdmin" "SuperAdmin" | has $crd.spec.accessLevel) }}
+      {{- cat "Unsupported accessLevel type" $crd.spec.accessLevel "in" $crd.name "crd" | fail }}
+    {{- end }}
+
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: user-authz:{{ $crd.name }}:{{ $crd.spec.accessLevel | kebabcase }}
```
Comment (apolovov) [inline ee/modules/140-user-authz/templates/multitenancy/role-bindings.yaml:12]: it is better to add d8: prefix!

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/140-user-authz/templates/multitenancy/role-bindings.yaml (hunk 10 above)
Comment (yalosev) [reply ee/modules/140-user-authz/templates/multitenancy/role-bindings.yaml:12]: Well, current module [doesn't contain](https://github.com/deckhouse/deckhouse/blob/main/modules/140-user-authz/templates/cluster-role-bindings.yaml#L8) this prefix. Are you sure, that we have to rename all of this?

File: modules/140-user-authz/hooks/handle_cluster_authorization_rules.go [hunk 11]
```diff
@@ -60,18 +50,87 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
...
+			multiRule := processMultitenancyCAR(input, car, allNamespaces)
+			multitenancyCARS = append(multitenancyCARS, multiRule)
+		} else {
+			simpleCARs = append(simpleCARs, car.ToValues())
+		}
 	}
 
-	input.Values.Set("userAuthz.internal.crds", ccrs)
+	input.Values.Set("userAuthz.internal.crds", simpleCARs)
+	input.Values.Set("userAuthz.internal.multitenancyCRDs", multitenancyCARS)
```
[the remark this answers is not in the document]
Comment (yalosev) [reply modules/140-user-authz/hooks/handle_cluster_authorization_rules.go:92]: It's a bit hard, to rename crds everywhere in templates. To keep this equal `multitenancyCRDs` was chosen

**Comment by EvgenySamoylov** [comment]
Closes #690

**Comment by yalosev** [comment]
Outdated

**Pull request closed without merging**